### PR TITLE
Machine Learning Services - Model detection in predict wizard

### PR DESCRIPTION
**Pull request by llali**
This PR includes 
- some refactoring and code cleaning to be able to add tests.
- unit tests for services and some UI
- extracting model inputs and outputs and displaying in UI
Screen Shot 2020-03-20 at 10 30 51 AM
Screen Shot 2020-03-20 at 10 31 35 AM

### CODE REVIEW

File: extensions/machine-learning-services/src/common/utils.ts [hunk 1]
```diff
@@ -227,3 +227,20 @@ export function getRegisteredModelsTowPartsName(config: Config) {
 	const tableName = doubleEscapeSingleBrackets(config.registeredModelTableName);
 	return `[${schema}].[${tableName}]`;
 }
+
+export function hexToBytes(hexContent: string): number[] {
```
Comment (llali) [inline extensions/machine-learning-services/src/common/utils.ts]: remove

File: extensions/machine-learning-services/src/common/utils.ts [hunk 2]
```diff
@@ -227,3 +227,20 @@ export function getRegisteredModelsTowPartsName(config: Config) {
...
+	let bytes = [];
+	if (hexContent) {
+		for (let c = 0; c < hexContent.length; c += 2) {
+			bytes.push(parseInt(hexContent.substr(c, 2), 16));
+		}
+	}
+	return bytes;
+}
+
+export async function writeFileFromHex(content: string): Promise<string> {
```
Comment (llali) [inline extensions/machine-learning-services/src/common/utils.ts]: add comment

File: extensions/machine-learning-services/src/modelManagement/registeredModelService.ts [hunk 3]
```diff
@@ -70,6 +70,26 @@ export class RegisteredModelService {
 		return updatedModel;
 	}
 
+	public async downloadModel(model: RegisteredModel): Promise<string> {
```
Comment (llali) [inline extensions/machine-learning-services/src/modelManagement/registeredModelService.ts]: comment

File: extensions/machine-learning-services/src/views/models/prediction/predictWizard.ts [hunk 4]
```diff
@@ -52,14 +51,14 @@ export class PredictWizard extends ModelViewBase {
...
-				wizard.backButton.enabled = false;
-				await this.predict();
-				wizard.cancelButton.enabled = true;
-				wizard.backButton.enabled = true;
-				if (this._parentView) {
-					this._parentView?.refresh();
+			if (validated) {
+				if (pageInfo.newPage === undefined) {
+					this.onLoading();
+					await this.predict();
```
Comment (llali) [inline extensions/machine-learning-services/src/views/models/prediction/predictWizard.ts]: cleanup downloaded models

File: extensions/machine-learning-services/src/views/models/prediction/inputColumnsComponent.ts [hunk 5]
```diff
@@ -8,21 +8,26 @@ import { ModelViewBase } from '../modelViewBase';
...
 	private _form: azdata.FormContainer | undefined;
 	private _databases: azdata.DropDownComponent | undefined;
 	private _tables: azdata.DropDownComponent | undefined;
 	private _columns: ColumnsTable | undefined;
 	private _dbNames: string[] = [];
 	private _tableNames: DatabaseTable[] = [];
+	private _modelParameters: ModelParameters | undefined;
+	private _dbTableComponent: azdata.FlexContainer | undefined;
 
+	//private componentLength = 200;
```
Comment (kisantia) [inline extensions/machine-learning-services/src/views/models/prediction/inputColumnsComponent.ts]: is this still needed?

**Comment by kisantia** [comment]
Did you also do some accessibility testing for this UI?

**Comment by llali** [comment]
verified some accessibility scenario but I haven't really test for any of the UIs in this extension. That would be a task for me for future when I release the extension

---
In reply to: [603450361](https://github.com/microsoft/azuredatastudio/pull/9609#issuecomment-603450361) [](ancestors = 603450361)

File: extensions/machine-learning-services/src/modelManagement/azureModelRegistryService.ts [hunk 6]
```diff
@@ -139,15 +153,14 @@ export class AzureModelRegistryService {
...
+				const client = await this.getAmlClient(account, subscription, tenant);
 				let result = resourceGroup ? await client.workspaces.listByResourceGroup(resourceGroup.name) : await client.workspaces.listBySubscription();
-				resources.push(...result);
+				if (result) {
+					resources.push(...result);
+				}
 			}
 		} catch (error) {
-
+			throw error;
```
Comment (Benjin) [inline extensions/machine-learning-services/src/modelManagement/azureModelRegistryService.ts]: Anything else supposed to happen in this catch block, or is it just for debugging?

File: extensions/machine-learning-services/src/modelManagement/azureModelRegistryService.ts [hunk 7]
```diff
@@ -263,16 +286,29 @@ export class AzureModelRegistryService {
...
+		account: azdata.Account,
+		subscription: azureResource.AzureResourceSubscription,
+		tenant: any,
+		options: AzureMachineLearningWorkspacesOptions | undefined = undefined,
+		apiVersion: string | undefined = undefined): Promise<AzureMachineLearningWorkspaces> {
+		if (this._amlClient) {
+			return this._amlClient;
+		} else {
+			const tokens = await this._apiWrapper.getSecurityToken(account, azdata.AzureResource.ResourceManagement);
+			const token = tokens[tenant.id].token;
```
Comment (Benjin) [inline extensions/machine-learning-services/src/modelManagement/azureModelRegistryService.ts]: Is it guaranteed that the tokens will have one with the specified tenant at this point?

File: extensions/machine-learning-services/src/modelManagement/deployedModelService.ts [hunk 8]
```diff
@@ -92,16 +104,40 @@ export class RegisteredModelService {
 			}
 		}
 	}
+	private loadModelData(row: azdata.DbCellValue[]): RegisteredModel {
+		return {
+			id: +row[0].displayValue,
```
Comment (Benjin) [inline extensions/machine-learning-services/src/modelManagement/deployedModelService.ts]: Is the `+` intentional?  (learning moment for me: if so, what does it do?)

File: extensions/machine-learning-services/src/modelManagement/deployedModelService.ts [hunk 9]
```diff
@@ -92,16 +104,40 @@ export class RegisteredModelService {
...
+		};
+	}
+
+	private async updateModel(model: RegisteredModel): Promise<RegisteredModel | undefined> {
+		let connection = await this.getCurrentConnection();
+		let updatedModel: RegisteredModel | undefined = undefined;
+		if (connection) {
+			const query = this.getUpdateModelQuery(connection.databaseName, model);
+			let result = await this._queryRunner.safeRunQuery(connection, query);
+			if (result && result.rows && result.rows.length > 0) {
```
Comment (Benjin) [inline extensions/machine-learning-services/src/modelManagement/deployedModelService.ts]: can the null-conditional operator be used here to simplify?

File: extensions/machine-learning-services/src/modelManagement/modelPythonClient.ts [hunk 10]
```diff
@@ -13,33 +13,89 @@ import * as utils from '../common/utils';
...
+			'import json',
+			`onnx_model_path = '${modelFolderPath}'`,
+			`onnx_model = onnx.load_model(onnx_model_path)`,
+			`type_map = {
+				onnx.TensorProto.DataType.FLOAT: 'real',
+				onnx.TensorProto.DataType.UINT8: 'tinyint',
+				onnx.TensorProto.DataType.INT16: 'smallint',
+				onnx.TensorProto.DataType.INT32: 'int',
+				onnx.TensorProto.DataType.INT64: 'bigint',
+				onnx.TensorProto.DataType.STRING: 'varchar(MAX)',
```
Comment (Benjin) [inline extensions/machine-learning-services/src/modelManagement/modelPythonClient.ts]: I haven't a clue about onnx, but figured I'd just check whether this ought to be nvarchar (does onnx support unicode/dualwidth character encoding?)

File: extensions/machine-learning-services/src/modelManagement/deployedModelService.ts (hunk 8 above)
Comment (llali) [reply extensions/machine-learning-services/src/modelManagement/deployedModelService.ts]: it convert to number

---
In reply to: [397442655](https://github.com/microsoft/azuredatastudio/pull/9609#discussion_r397442655) [](ancestors = 397442655)

File: extensions/machine-learning-services/src/modelManagement/deployedModelService.ts (hunk 9 above)
Comment (llali) [reply extensions/machine-learning-services/src/modelManagement/deployedModelService.ts]: you mean result?.rows?  I can replace that but still have to check for length

---
In reply to: [397443949](https://github.com/microsoft/azuredatastudio/pull/9609#discussion_r397443949) [](ancestors = 397443949)

File: extensions/machine-learning-services/src/modelManagement/modelPythonClient.ts (hunk 10 above)
Comment (llali) [reply extensions/machine-learning-services/src/modelManagement/modelPythonClient.ts]: hmm not sure, got the mapping from the service team. will  double check 

---
In reply to: [397444805](https://github.com/microsoft/azuredatastudio/pull/9609#discussion_r397444805) [](ancestors = 397444805)

**Pull request merged**